### PR TITLE
Update vehicle object struct

### PR DESCRIPTION
**Pull request by duncanspumpkin**
Information from the vehicle update branch. Consider further rework on names

### CODE REVIEW

**Comment by duncanspumpkin** [comment]
I'm not really happy with the names I chose for the vehicle_type and vehicle_class fields. I saw someone had named it vehicle_category in one note. I'm open for suggestions.

File: src/openloco/things/vehicle.h [hunk 1]
```diff
@@ -26,7 +26,7 @@ namespace openloco
         vehicle_6,
     };
 
-    // Unrelated to vehicle_thing_type -- we should probably consider renaming when things become clearer.
+    // Related to vehicle_type
```
Comment (AaronVanGeffen) [inline src/openloco/things/vehicle.h]: Might as well make the count part of the enum now, I guess. :)

File: src/openloco/objects/vehicle_object.h [hunk 2]
```diff
@@ -4,6 +4,24 @@
 
 namespace openloco
 {
+    enum class vehicle_class : uint8_t
```
Comment (AaronVanGeffen) [inline src/openloco/objects/vehicle_object.h]: vehicle_medium?

**Comment by AaronVanGeffen** [comment]
Perhaps we could rename what is now `vehicle_thing_type` to `vehice_part_type`?

File: src/openloco/things/vehicle.h [hunk 3]
```diff
@@ -88,7 +88,7 @@ namespace openloco
         thing_id_t next_car_id;     // 0x3A
         uint8_t pad_3C[0x40 - 0x3C];
         uint16_t object_id; // 0x40
-        uint8_t var_42;
+        vehicle_class v_class;
```
Comment (AaronVanGeffen) [inline src/openloco/things/vehicle.h]: As this now has a proper type, the loop constructs in `vehicles_menu_mouse_down` and `build_vehicles_menu_mouse_down` could use it as well:

https://github.com/OpenLoco/OpenLoco/blob/0a738d35ab68903ded48739e0b5eeacfd8033c7e/src/openloco/windows/toolbar_top.cpp#L646

https://github.com/OpenLoco/OpenLoco/blob/0a738d35ab68903ded48739e0b5eeacfd8033c7e/src/openloco/windows/toolbar_top.cpp#L720

Comment (duncanspumpkin) [reply src/openloco/things/vehicle.h]: Yes sorry that was my plan. I just didn't have the time this evening.

Comment (AaronVanGeffen) [reply src/openloco/things/vehicle.h]: Oh it's totally fine, I'd expected you to say it'd be outside of the scope of this PR, even. :) Just wanted to note that it opened up the possibility.

File: src/openloco/objects/vehicle_object.h (hunk 2 above)
Comment (marijnvdwerf) [reply src/openloco/objects/vehicle_object.h]: `TransportMode`?

Comment (IntelOrca) [reply src/openloco/objects/vehicle_object.h]: `VehicleKind`?

Comment (AaronVanGeffen) [reply src/openloco/objects/vehicle_object.h]: `TransportMode` sounds fine to me, even if it doesn't sound vehicle-exclusive. To me, `VehicleKind` sounds too close to `VehicleType`, though, which we already have.

Comment (duncanspumpkin) [reply src/openloco/objects/vehicle_object.h]: Yeah I think TransportMode sounds the best. I'll go for that.

File: src/openloco/things/vehicle.h (hunk 3 above)
Comment (duncanspumpkin) [reply src/openloco/things/vehicle.h]: The next question is how best to implement the array. Do i make a map that looks up vehicle type or do I just go for an array and cast the vehicle type to an int.

Comment (AaronVanGeffen) [reply src/openloco/things/vehicle.h]: If you make it an `enum` rather than an `enum class` the type should implicitly convert to `uint8_t`.

Comment (duncanspumpkin) [reply src/openloco/things/vehicle.h]: I think strongly typed is good for all the other uses of it though. I think I'll just do a cast.

Comment (AaronVanGeffen) [reply src/openloco/things/vehicle.h]: `static_cast` should be fine too, yeah.

Comment (marijnvdwerf) [reply src/openloco/things/vehicle.h]: If we're going for strongly typed, wouldn't a map make more sense?

File: src/openloco/windows/toolbar_top.cpp [hunk 4]
```diff
@@ -633,6 +634,23 @@ namespace openloco::ui::windows::toolbar_top
...
 
+    struct VehicleTypeInterfaceParam
+    {
+        uint32_t image;
+        string_id build_string;
+        string_id num_singular;
+        string_id num_plural;
+    };
+
+    const std::map<VehicleType, VehicleTypeInterfaceParam> VehicleTypeInterfaceParameters{
```
Comment (AaronVanGeffen) [inline src/openloco/windows/toolbar_top.cpp]: `static const`

File: src/openloco/windows/toolbar_top.cpp [hunk 5]
```diff
@@ -717,48 +717,22 @@ namespace openloco::ui::windows::toolbar_top
...
-                string_ids::num_trains_plural,
-                string_ids::num_buses_plural,
-                string_ids::num_trucks_plural,
-                string_ids::num_trams_plural,
-                string_ids::num_aircraft_plural,
-                string_ids::num_ships_plural,
-            };
+            auto& interface_param = VehicleTypeInterfaceParameters.at(static_cast<VehicleType>(vehicleType));
 
+            uint32_t vehicle_image = gfx::recolour(interface_param.image, company_colour);
```
Comment (AaronVanGeffen) [inline src/openloco/windows/toolbar_top.cpp]: You've replaced the vehicle image id with the build vehicle image id.

Comment (duncanspumpkin) [reply src/openloco/windows/toolbar_top.cpp]: Doh

File: src/openloco/objects/vehicle_object.h [hunk 6]
```diff
@@ -4,6 +4,25 @@
...
+
+    enum class VehicleType : uint8_t
+    {
+        train = 0,
+        bus,
+        truck,
+        tram,
+        plane,
+        ship,
+        Count
```
Comment (marijnvdwerf) [inline src/openloco/objects/vehicle_object.h]: Is adding `count` a good idea? It'll mean that switch statements will automatically have a `case VehicleType::count`: added. I'd rather see a `vehicleTypeCount` constant, or a list that can be foreached over.

Comment (AaronVanGeffen) [reply src/openloco/objects/vehicle_object.h]: I suggested this in the other PR, but concede it's not the best idea, in retrospect.

Comment (duncanspumpkin) [reply src/openloco/objects/vehicle_object.h]: I was just following what was done in openrct2. I can change it to a constexpr if that is preferred.

**Comment by duncanspumpkin** [comment]
`vehicle_thing_type` I don't really want to change that at this moment before we better understand the things.

**Pull request merged**